### PR TITLE
[Roosterteeth] Update extractor to support new site.

### PR DESCRIPTION
**Pull request by ddmgy**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This updates the extractor for Roosterteeth.com to support the recent changes to the website. It now uses the officially provided API rather than scraping HTML. Addresses issue #16094.

There are some issues that need to be fixed before it can be merged.

- ~~Videos that do not require log in can be downloaded with no issue, but attempting to log in fails. I have not yet found a way to log in through the new API. The old log in URL stills exists, and can be used to log in through a browser. However, the form is now loaded through Javascript after the page loads, so youtube-dl seems unable to grab it.~~ Logging in and downloading FIRST-only videos now works.

- ~~The changes I made don't look great, as I just wanted to get it mostly working before hopefully getting feedback on the log in issue. I'll be going through and fixing things up over the weekend.~~ I've cleaned things up and made user-facing strings less generic. I'm open to any suggestions for further changes.

Logging in is done by POSTing username and password to an authorization URL, and an access token is returned. ~~This access token is a cookie in the browser, so it could be saved as a cookie here, but I'm not sure how to do that, or how to check if a saved cookie is expired. I'll look more in to it.~~

Barring any requested changes, this seems ready to be merged.

Edit: The access token is a standard JWT that is valid for 8 hours after logging in. I have implemented saving the access token as a cookie and re-downloading it after it expires, so users who download multiple videos in a short time will not have to log in every time.

### CODE REVIEW

**Comment by DIzFer** [comment]
I came here looking for this precise fix and I can confirm it works for me. Is it usual for a PR like this to sit unmerged for 3 weeks?

**Comment by au5ton** [comment]
bump because this is still unmerged on the master branch

File: youtube_dl/extractor/roosterteeth.py [hunk 1]
```diff
@@ -54,86 +54,110 @@ def _login(self):
...
 
+        created_at = response.get('created_at', 0)
+        expires_in = response.get('expires_in', 0)
+
+        self._set_cookie(
+            '.roosterteeth.com',
+            'rt_access_token',
+            self._ACCESS_TOKEN,
+            created_at + expires_in
+        )
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: Remove excessive verbosity. This fits well on a single line.

File: youtube_dl/extractor/roosterteeth.py [hunk 2]
```diff
@@ -54,86 +54,110 @@ def _login(self):
...
-        title = strip_or_none(self._og_search_title(
-            webpage, default=None)) or episode
-
-        m3u8_url = self._search_regex(
-            r'file\s*:\s*(["\'])(?P<url>http.+?\.m3u8.*?)\1',
-            webpage, 'm3u8 url', default=None, group='url')
-
+        headers = {}
+        if self._ACCESS_TOKEN:
+            headers['Authorization'] = 'Bearer {}'.format(self._ACCESS_TOKEN)
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: No `{}`.

File: youtube_dl/extractor/roosterteeth.py [hunk 3]
```diff
@@ -54,86 +54,110 @@ def _login(self):
...
+
+        api_response = self._call_api(
+            display_id,
+            note='Downloading video information (1/2)',
+            errnote='Unable to download video information (1/2)',
+            headers=headers,
+        )
+
+        if len(api_response.get('data', [])) == 0:
+            raise ExtractorError('Unable to download video information')
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: 105-106 remove.

File: youtube_dl/extractor/roosterteeth.py [hunk 4]
```diff
@@ -54,86 +54,110 @@ def _login(self):
...
+        data = api_response.get('data')[0]
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: Mandatory data must be accessed with `[]` not `get`.

File: youtube_dl/extractor/roosterteeth.py [hunk 5]
```diff
@@ -54,86 +54,110 @@ def _login(self):
...
+
+        attributes = data.get('attributes', {})
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: Mandatory.

File: youtube_dl/extractor/roosterteeth.py [hunk 6]
```diff
@@ -54,86 +54,110 @@ def _login(self):
...
+        episode = attributes.get('display_title')
+        title = attributes.get('title')
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: Mandatory.

File: youtube_dl/extractor/roosterteeth.py [hunk 7]
```diff
@@ -54,86 +54,110 @@ def _login(self):
...
+        description = attributes.get('caption')
+        series = attributes.get('show_title')
+
+        images = data.get('included', {}).get('images')
+        if images and len(images) > 0:
+            images = images[0]
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: `images` is not guaranteed to be a list.

File: youtube_dl/extractor/roosterteeth.py [hunk 8]
```diff
@@ -54,86 +54,110 @@ def _login(self):
...
+        data = api_response.get('data')[0]
+
+        attributes = data.get('attributes', {})
+        episode = attributes.get('display_title')
+        title = attributes.get('title')
+        description = attributes.get('caption')
+        series = attributes.get('show_title')
+
+        images = data.get('included', {}).get('images')
+        if images and len(images) > 0:
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: `if images` implies the second check.

File: youtube_dl/extractor/roosterteeth.py [hunk 9]
```diff
@@ -54,86 +54,110 @@ def _login(self):
...
+                if now < member_golive:
+                    self._golive_error(display_id, 'site members')
+                elif now < public_golive:
+                    self._golive_error(display_id, 'the public')
+                else:
+                    raise ExtractorError('Video is not available')
+
+        if len(video_response.get('data', [])) == 0:
+            raise ExtractorError('Unable to download video information')
+        video_attributes = video_response.get('data')[0].get('attributes')
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: Same.

File: youtube_dl/extractor/roosterteeth.py [hunk 10]
```diff
@@ -143,6 +167,25 @@ def _real_extract(self, url):
...
+    def _call_api(self, video_id, path=None, **kwargs):
+        url = self._API_URL + video_id
+        if path:
+            url = url + path
+
+        return self._download_json(
+            url,
+            video_id,
+            **kwargs
+        )
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: Single line.

File: youtube_dl/extractor/roosterteeth.py [hunk 11]
```diff
@@ -54,86 +54,110 @@ def _login(self):
...
+            raise ExtractorError('Unable to download video information')
+        data = api_response.get('data')[0]
+
+        attributes = data.get('attributes', {})
+        episode = attributes.get('display_title')
+        title = attributes.get('title')
+        description = attributes.get('caption')
+        series = attributes.get('show_title')
+
+        images = data.get('included', {}).get('images')
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: `try_get`.

File: youtube_dl/extractor/roosterteeth.py [hunk 12]
```diff
@@ -54,86 +54,110 @@ def _login(self):
...
+        if images and len(images) > 0:
+            images = images[0]
+        thumbnail = images.get('attributes', {}).get('thumb')
```
Comment (dstftw) [inline youtube_dl/extractor/roosterteeth.py]: `try_get`.

File: youtube_dl/extractor/roosterteeth.py (hunk 9 above)
Comment (ddmgy) [reply youtube_dl/extractor/roosterteeth.py]: I'm not sure what 'Same' is meant to be referencing, but I removed the check and error, and changed video_attributes declaration to use try_get. If this is not what you meant for me to change, I can revert it.

**Comment by ddmgy** [comment]
I've (hopefully) made the requested changes. Awaiting further review.

**Comment by RichardHancock** [comment]
Thanks for this, seems to be working on every video I've tried so far. One critique would be that the thumbnails it fetches are tiny (240ish), even though there are 720-1080p thumbnails available on the video page, minor thing but would be great if someone could fix. 
Thanks

**Comment by ddmgy** [comment]
@RichardHancock It should now download a larger image for the thumbnail.

**Comment by RichardHancock** [comment]
@ddmgy Thanks that works great. Seems to download with incorrect image file extension for some videos, but that's Rooster Teeth's end as I had this issue when manually downloading them. 

Thanks for doing this, works just as good as the old website extractor now, Hopefully someone can get this merged in soon

**Comment by ddmgy** [comment]
@RichardHancock Glad to hear it works. Would you mind telling me which videos are giving the wrong extension for thumbnails? I'll see if I can fix it.

**Comment by RichardHancock** [comment]
@ddmgy Seems to be pretty random, only ones I can find doing it right now are recent Camp Camp episodes. I assume one departments just uploading them wrong. 
Here's one that I've checked: https://roosterteeth.com/episode/camp-camp-season-3-7 
That one downloads a .png but when opened in some programs they throw an error saying that its actually a jpg (other programs just silently load it anyway). 
It doesn't seem to be as bad as when I was manually downloading the thumbnails as some of those images were .webp with .png extensions which almost no program could detect and fix. I just assumed it was the same problem.

I don't really think it's anything that you need to fix, it's just their mistake. I'll post again if I find other series doing it.
Also that latest commit you made makes it download the small thumbnails again, but I assume your just testing things.

Thanks

**Comment by ddmgy** [comment]
@RichardHancock Hmm, the thumbnails returned for that [Camp Camp episode](https://svod-be.roosterteeth.com/api/v1/episodes/camp-camp-season-3-7) are a mix of PNGs and JPGs, and I'm not sure which set of those youtube-dl is actually grabbing. ~~I would assume the API is returning the data in the same order every time, so it should be getting PNGs, but I'll test that more tonight.~~ Actually, I see now that even when the extension on the URL shows it is a PNG, it is sometimes actually a JPG image. The image opens fine for me in everything I'm using, but you're right: this is on RoosterTeeth's side, and there's nothing I can fix there.

~~Unfortunately, even if I always manage to grab the same set of thumbnails for a video, there is no way to tell what size each of the images is, and they don't always seem to be ordered correctly. When --write-thumbnail is passed to youtube-dl, it will [only download the last thumbnail](https://github.com/rg3/youtube-dl/blob/master/youtube_dl/YoutubeDL.py#L2358) in the list of thumbnails (which is what my latest commit generates). You can use --write-all-thumbnails, which will download all 4 thumbnails, but then you'd have to check each one manually to see which is the size you want and then delete the other 3.~~

Well, you can still use --write-all-thumbnails and manually select which image you want to keep, but using --write-thumbnail should at least choose to download only the largest available thumbnail now. I wish there was a proper way to choose which thumbnail you want to download, but I'm not seeing any available options for that.

Thanks for your feedback on this. I don't use thumbnails, so it probably never would have come to my attention that I was grabbing the wrong size.

**Comment by ddmgy** [comment]
As I no longer have a FIRST subscription with roosterteeth.com, I will be unable to make any further changes to this pull request, in the unlikely event that it is reviewed again or considered for merging.

If anyone wants to take over maintenance of this pull request, leave a comment here so we can figure out a way to transfer it over.

**Comment by RichardHancock** [comment]
@ddmgy I would be willing to take it over if needed.

**Comment by RichardHancock** [comment]
@ddmgy Also just realized that the changes weren't marked as resolved so I just did that, hopefully that will trigger a re-review of the pull request.

**Comment by RichardHancock** [comment]
The test for the extractor is failing at the moment, 
`'episode': 'S2:E10 - Million Dollars, But... The Game Announcement'`
Needs to change to:
`'episode': 'Million Dollars, But... The Game Announcement'`
for it to pass the test

**Comment by lanodan** [comment]
Hi, any ways this could support bonus episodes? For example https://roosterteeth.com/bonus-feature/rwby-adam-short (volume 6, adam character)

**Comment by AevumDecessus** [comment]
@dstftw 
I've updated the tests for this change, and they're passing now. I can confirm that the module is working with the new Rooster Teeth site, both for public videos, as well as ones that require a FIRST subscription/login in order to download.

**Comment by mDuo13** [comment]
Any updates to this or the other PR for presumably the same problem (#17843)? It's pretty inconvenient using the API workaround described in #16094, so I'd encourage @dstftw or other maintainers to merge this...

Given that the Rooster Teeth extractor doesn't work at all in the current master branch, I would think there's not as much worry about breaking it with changes like these...

**Comment by lanodan** [comment]
[2019-03-10 00:31:52] Rome Reginelli:
> Any updates to this or the other PR for presumably the same problem (#17843)? It's pretty inconvenient using the API workaround described in #16094, so I'd encourage @dstftw or other maintainers to merge this...
> 
> Given that the Rooster Teeth extractor doesn't work at all in the current master branch, I would think there's not as much worry about breaking it with changes like these...

Not much to break and this PR works, I have it on my youtube-dl installation (thanks to gentoo allowing to just drop a patchfile) for quite a while.

**Comment by tylermauthe** [comment]
FWIW I used this change to download stuff... so, it does work... It'd be really nice to just merge this..................... .. .. . . . . . . . . . . . . . .

**Comment by tylermauthe** [comment]
The requested changes seem pretty pedantic too - just merge it 🚀

**Comment by Qyriad** [comment]
Weren't the requested changes *made* anyway? Or at least, it doesn't seem to me like there's been any review since https://github.com/ytdl-org/youtube-dl/pull/16105#issuecomment-396768722.
I don't really understand why neither this nor #17843 have been merged, given that the Roosterteeth extractor is completely non-functional on master. 
I've had this merged on my local installation for months.

**Comment by au5ton** [comment]
Please merge.

**Comment by jonnyrobbie** [comment]
I don't know - I'm still having problems:

```
./bin/youtube-dl --verbose https://roosterteeth.com/episode/rwby-volume-6-1
[debug] System config: []
[debug] User config: ['--no-mtime', '--write-sub', '--sub-format', 'best', '--embed-subs', '--add-metadata', '--xattrs', '--mark-watched']
[debug] Custom config: []
[debug] Command-line args: ['--verbose', 'https://roosterteeth.com/episode/rwby-volume-6-1']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2019.04.24
[debug] Python version 3.7.3 (CPython) - Linux-5.0.13-arch1-1-ARCH-x86_64-with-arch
[debug] exe versions: ffmpeg 4.1.3, ffprobe 4.1.3
[debug] Proxy map: {}
[RoosterTeeth] rwby-volume-6-1: Downloading webpage
ERROR: Unable to extract m3u8 URL; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/youtube_dl/YoutubeDL.py", line 796, in extract_info
    ie_result = ie.extract(url)
  File "/usr/lib/python3.7/site-packages/youtube_dl/extractor/common.py", line 529, in extract
    ie_result = self._real_extract(url)
  File "/usr/lib/python3.7/site-packages/youtube_dl/extractor/roosterteeth.py", line 115, in _real_extract
    raise ExtractorError('Unable to extract m3u8 URL')
youtube_dl.utils.ExtractorError: Unable to extract m3u8 URL; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
```

**Comment by lanodan** [comment]
[2019-05-09 12:41:13-0700] jonnyrobbie:
> I don't know - I'm still having problems:
> 
> ```
> ./bin/youtube-dl --verbose https://roosterteeth.com/episode/rwby-volume-6-1
> [debug] System config: []
> [debug] User config: ['--no-mtime', '--write-sub', '--sub-format', 'best', '--embed-subs', '--add-metadata', '--xattrs', '--mark-watched']
> [debug] Custom config: []
> [debug] Command-line args: ['--verbose', 'https://roosterteeth.com/episode/rwby-volume-6-1']
> [debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
> [debug] youtube-dl version 2019.04.24
> [debug] Python version 3.7.3 (CPython) - Linux-5.0.13-arch1-1-ARCH-x86_64-with-arch
> [debug] exe versions: ffmpeg 4.1.3, ffprobe 4.1.3
> [debug] Proxy map: {}
> [RoosterTeeth] rwby-volume-6-1: Downloading webpage
> ERROR: Unable to extract m3u8 URL; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
> Traceback (most recent call last):
>   File "/usr/lib/python3.7/site-packages/youtube_dl/YoutubeDL.py", line 796, in extract_info
>     ie_result = ie.extract(url)
>   File "/usr/lib/python3.7/site-packages/youtube_dl/extractor/common.py", line 529, in extract
>     ie_result = self._real_extract(url)
>   File "/usr/lib/python3.7/site-packages/youtube_dl/extractor/roosterteeth.py", line 115, in _real_extract
>     raise ExtractorError('Unable to extract m3u8 URL')
> youtube_dl.utils.ExtractorError: Unable to extract m3u8 URL; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
> ```

Weird, here it just works:

```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['--verbose', 'https://roosterteeth.com/episode/rwby-volume-6-1']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2019.04.30
[debug] Python version 3.6.5 (CPython) - Linux-4.19.27-gentoo-r1-x86_64-AMD_FX-tm-8350_Eight-Core_Processor-with-gentoo-2.6
[debug] exe versions: ffmpeg 4.1.1, ffprobe 4.1.1, rtmpdump 2.4
[debug] Proxy map: {}
[RoosterTeeth] rwby-volume-6-1: Downloading video information (1/2)
[RoosterTeeth] rwby-volume-6-1: Downloading video information (2/2)
[RoosterTeeth] rwby-volume-6-1: Downloading m3u8 information
[debug] Default format spec: bestvideo+bestaudio/best
[debug] Invoking downloader on 'https://rtv3-roosterteeth.akamaized.net/store/777dcdfb41acc2cebc48a69a94ab4460-f6785b63/ts/f6785b63-hls_1080p-store-777dcdfb41acc2cebc48a69a94ab4460.m3u8'
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 339
[download] Destination: Argus Limited-29129.mp4
[download] 100% of 864.13MiB in 02:40
[debug] ffmpeg command line: ffprobe -show_streams 'file:Argus Limited-29129.mp4'
[ffmpeg] Fixing malformed AAC bitstream in "Argus Limited-29129.mp4"
[debug] ffmpeg command line: ffmpeg -y -loglevel repeat+info -i 'file:Argus Limited-29129.mp4' -c copy -f mp4 -bsf:a aac_adtstoasc 'file:Argus Limited-29129.temp.mp4'
```

The patch I picked up for it is available at (I didn't modify it but
the code in the PR might have changed in the meantime):
https://hacktivis.me/git/etc_portage.work/patches/net-misc/youtube-dl/8b703800ecb-pull-16105-rooster-teeth.diff

**Comment by jonnyrobbie** [comment]
> Weird, here it just works:

yeah, that was my bad. I cloned the repos and switched the branches, and then just run the bin/ executable, which just imports the system vanilla version of ytdl. I should have run it with `python -m youtube_dl`.

**Comment by axipher** [comment]
Any update on if this is getting merged in to the official build?  Or if it still works?

**Comment by jonnyrobbie** [comment]
I've tried that a few days ago and it still worked for me. But I was simply downloading single episodes by the simple episode url. Maybe some more advanced querries still cause issues? I don't know.

**Comment by JordanCarr** [comment]
The URL scheme for Rooster Teeth appears to have changed from roosterteeth.com/episode/foo to roosterteeth.com/watch/foo. When using one of the new /watch/ URLs the extractor doesn’t recognize it for the RoosterTeeth extractor and then the generic one fails. Even with the new URL simply changing /watch/ to /episode/ allows it to continue to work.

**Comment by agent619** [comment]
I'm trying to use this version and I'm getting a 403 error:

youtube-dl --verbose https://roosterteeth.com/episode/backwardz-compatible-2019-sadism-or-masochism
[debug] System config: []
[debug] User config: [u'-o', u'~/Movies/%(title)s.%(ext)s', u'-f', u'bestvideo[height<=640]+bestaudio[height<=640]/best[height<=640]']
[debug] Custom config: []
[debug] Command-line args: [u'--verbose', u'https://roosterteeth.com/episode/backwardz-compatible-2019-sadism-or-masochism']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2019.08.02
[debug] Python version 2.7.11 (CPython) - Darwin-18.6.0-x86_64-i386-64bit
[debug] exe versions: ffmpeg 4.1.4, ffprobe 4.1.4, rtmpdump 2.4
[debug] Proxy map: {}
[RoosterTeeth] backwardz-compatible-2019-sadism-or-masochism: Downloading video information (1/2)
[RoosterTeeth] backwardz-compatible-2019-sadism-or-masochism: Downloading video information (2/2)
ERROR: Unable to download video information (2/2): HTTP Error 403: Forbidden (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/youtube_dl-2019.8.2-py2.7.egg/youtube_dl/extractor/common.py", line 627, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/youtube_dl-2019.8.2-py2.7.egg/youtube_dl/YoutubeDL.py", line 2227, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 437, in open
    response = meth(req, response)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 550, in http_response
    'http', request, response, code, msg, hdrs)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 475, in error
    return self._call_chain(*args)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 409, in _call_chain
    result = func(*args)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 558, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)

**Comment by JordanCarr** [comment]
> I'm trying to use this version and I'm getting a 403 error:
> 
> youtube-dl --verbose https://roosterteeth.com/episode/backwardz-compatible-2019-sadism-or-masochism
> [debug] System config: []
> [debug] User config: [u'-o', u'~/Movies/%(title)s.%(ext)s', u'-f', u'bestvideo[height<=640]+bestaudio[height<=640]/best[height<=640]']
> [debug] Custom config: []
> [debug] Command-line args: [u'--verbose', u'https://roosterteeth.com/episode/backwardz-compatible-2019-sadism-or-masochism']
> [debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
> [debug] youtube-dl version 2019.08.02
> [debug] Python version 2.7.11 (CPython) - Darwin-18.6.0-x86_64-i386-64bit
> [debug] exe versions: ffmpeg 4.1.4, ffprobe 4.1.4, rtmpdump 2.4
> [debug] Proxy map: {}
> [RoosterTeeth] backwardz-compatible-2019-sadism-or-masochism: Downloading video information (1/2)
> [RoosterTeeth] backwardz-compatible-2019-sadism-or-masochism: Downloading video information (2/2)
> ERROR: Unable to download video information (2/2): HTTP Error 403: Forbidden (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see https://yt-dl.org/update on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
> File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/youtube_dl-2019.8.2-py2.7.egg/youtube_dl/extractor/common.py", line 627, in _request_webpage
> return self._downloader.urlopen(url_or_request)
> File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/youtube_dl-2019.8.2-py2.7.egg/youtube_dl/YoutubeDL.py", line 2227, in urlopen
> return self._opener.open(req, timeout=self._socket_timeout)
> File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 437, in open
> response = meth(req, response)
> File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 550, in http_response
> 'http', request, response, code, msg, hdrs)
> File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 475, in error
> return self._call_chain(*args)
> File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 409, in _call_chain
> result = func(*args)
> File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 558, in http_error_default
> raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)

It looks like that video is behind the FIRST subscription and you’ve not supplied a login so as to access FIRST content. Try supplying a username with --username and entering your password when prompted. You could also use a cookie file if the previous method doesn’t work

**Comment by agent619** [comment]
> > I'm trying to use this version and I'm getting a 403 error:
> > youtube-dl --verbose https://roosterteeth.com/episode/backwardz-compatible-2019-sadism-or-masochism
> > [debug] System config: []
> > [debug] User config: [u'-o', u'~/Movies/%(title)s.%(ext)s', u'-f', u'bestvideo[height<=640]+bestaudio[height<=640]/best[height<=640]']
> > [debug] Custom config: []
> > [debug] Command-line args: [u'--verbose', u'https://roosterteeth.com/episode/backwardz-compatible-2019-sadism-or-masochism']
> > [debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
> > [debug] youtube-dl version 2019.08.02
> > [debug] Python version 2.7.11 (CPython) - Darwin-18.6.0-x86_64-i386-64bit
> > [debug] exe versions: ffmpeg 4.1.4, ffprobe 4.1.4, rtmpdump 2.4
> > [debug] Proxy map: {}
> > [RoosterTeeth] backwardz-compatible-2019-sadism-or-masochism: Downloading video information (1/2)
> > [RoosterTeeth] backwardz-compatible-2019-sadism-or-masochism: Downloading video information (2/2)
> > ERROR: Unable to download video information (2/2): HTTP Error 403: Forbidden (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see https://yt-dl.org/update on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
> > File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/youtube_dl-2019.8.2-py2.7.egg/youtube_dl/extractor/common.py", line 627, in _request_webpage
> > return self._downloader.urlopen(url_or_request)
> > File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/youtube_dl-2019.8.2-py2.7.egg/youtube_dl/YoutubeDL.py", line 2227, in urlopen
> > return self._opener.open(req, timeout=self._socket_timeout)
> > File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 437, in open
> > response = meth(req, response)
> > File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 550, in http_response
> > 'http', request, response, code, msg, hdrs)
> > File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 475, in error
> > return self._call_chain(*args)
> > File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 409, in _call_chain
> > result = func(*args)
> > File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 558, in http_error_default
> > raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
> 
> It looks like that video is behind the FIRST subscription and you’ve not supplied a login so as to access FIRST content. Try supplying a username with --username and entering your password when prompted. You could also use a cookie file if the previous method doesn’t work

Ah I see! Thanks for that. I have clearly misunderstood ddmgy's description - I had thought that her extractor meant that I didn't have to log in at all. I'll try this then!

**Comment by lanodan** [comment]
It has been fixed by c72dc20d099bbe1dc4ede83e8f94a7bc42d81532 and 5efbc1366f4e4d9d4969cbfb404657349a5b3f99 , so this Pull Request can probably be closed with https://github.com/ytdl-org/youtube-dl/issues/16094 or should be rebased on current master if there is still missing pieces to be fixed (maybe the required accounts stuff).

**Pull request closed without merging**